samza-api/src/main/java/org/apache/samza/operators/StreamGraph.java

```diff
@@ -0,0 +1,94 @@
+ */
+package org.apache.samza.operators;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.data.MessageEnvelope;
+import org.apache.samza.serializers.Serde;
+
+import java.util.Map;
+
+
+/**
+ * Job-level programming interface to create an operator DAG and run in various different runtime environments.
+ */
+@InterfaceStability.Unstable
+public interface StreamGraph {
+  /**
+   * Method to add an input {@link MessageStream} from the system
+   *
+   * @param streamSpec  the {@link StreamSpec} describing the physical characteristics of the input {@link MessageStream}
+   * @param keySerde  the serde used to serialize/deserialize the message key from the input {@link MessageStream}
+   * @param msgSerde  the serde used to serialize/deserialize the message body from the input {@link MessageStream}
+   * @param <K>  the type of key in the input message
+   * @param <V>  the type of message in the input message
+   * @param <M>  the type of {@link MessageEnvelope} in the input {@link MessageStream}
+   * @return   the input {@link MessageStream} object
+   */
+  <K, V, M extends MessageEnvelope<K, V>> MessageStream<M> createInStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde);
+
+  /**
+   * Method to add an output {@link MessageStream} from the system
+   *
+   * @param streamSpec  the {@link StreamSpec} describing the physical characteristics of the output {@link MessageStream}
+   * @param keySerde  the serde used to serialize/deserialize the message key from the output {@link MessageStream}
+   * @param msgSerde  the serde used to serialize/deserialize the message body from the output {@link MessageStream}
+   * @param <K>  the type of key in the output message
+   * @param <V>  the type of message in the output message
+   * @param <M>  the type of {@link MessageEnvelope} in the output {@link MessageStream}
+   * @return   the output {@link MessageStream} object
+   */
+  <K, V, M extends MessageEnvelope<K, V>> OutputStream<M> createOutStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde);
+
+  /**
+   * Method to add an intermediate {@link MessageStream} from the system
+   *
+   * @param streamSpec  the {@link StreamSpec} describing the physical characteristics of the intermediate {@link MessageStream}
+   * @param keySerde  the serde used to serialize/deserialize the message key from the intermediate {@link MessageStream}
+   * @param msgSerde  the serde used to serialize/deserialize the message body from the intermediate {@link MessageStream}
+   * @param <K>  the type of key in the intermediate message
+   * @param <V>  the type of message in the intermediate message
+   * @param <M>  the type of {@link MessageEnvelope} in the intermediate {@link MessageStream}
+   * @return   the intermediate {@link MessageStream} object
+   */
+  <K, V, M extends MessageEnvelope<K, V>> OutputStream<M> createIntStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde);
```

Comment:
Definitely spell this out. It's easy to mix up with createInStream and if you do read it correctly the name *suggests* we're creating a stream of ints.